File: app/src/main/java/io/euphoria/xkcd/app/AboutActivity.java
package io.euphoria.xkcd.app;

import android.app.Activity;
import android.app.DownloadManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import io.euphoria.xkcd.app.updates.Manifest;
import io.euphoria.xkcd.app.updates.ManifestDownloader;
import io.euphoria.xkcd.app.updates.Version;

public class AboutActivity extends Activity {

    private enum UpdateCheckState { BLANK, BUSY, FAILED, NO_RESULT, UPDATE_AVAILABLE }

    private static final String UPDATE_CHECK_TAG = "UpdateCheck";

    private UpdateCheckState updateCheckState;
    private Manifest updateCheckResult;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.activity_about);

        updateCheckState = UpdateCheckState.BLANK;

        TextView version = findViewById(R.id.about_version);
        version.setText(getResources().getString(R.string.version_line,
                BuildConfig.VERSION_NAME,
                getResources().getString(R.string.app_version_label)));
    }

    private void setUpdateCheckState(UpdateCheckState newState) {
        updateCheckState = newState;
        Version newestVersion;
        if (updateCheckResult != null) {
            newestVersion = updateCheckResult.getVersion();
        } else {
            newestVersion = null;
        }
        Button button = findViewById(R.id.about_updateCheck);
        ViewGroup outputLine = findViewById(R.id.about_updateCheckLine);
        ProgressBar progress = findViewById(R.id.about_updateCheckProgress);
        TextView output = findViewById(R.id.about_updateCheckResult);
        switch (newState) {
            case BLANK:
                button.setEnabled(true);
                outputLine.setVisibility(View.GONE);
                progress.setVisibility(View.GONE);
                break;
            case BUSY:
                button.setEnabled(false);
                outputLine.setVisibility(View.VISIBLE);
                progress.setVisibility(View.VISIBLE);
                break;
            case FAILED: case NO_RESULT: case UPDATE_AVAILABLE:
                button.setEnabled(true);
                outputLine.setVisibility(View.VISIBLE);
                progress.setVisibility(View.GONE);
                break;
        }
        if (newState == UpdateCheckState.UPDATE_AVAILABLE) {
            button.setText(getResources().getString(R.string.update_check_download, newestVersion));
        } else {
            button.setText(R.string.update_check);
        }
        switch (newState) {
            case BUSY:
                output.setText(R.string.update_check_busy);
                break;
            case FAILED:
                output.setText(R.string.update_check_failed);
                break;
            case NO_RESULT:
                output.setText(R.string.update_check_no_result);
                break;
            case UPDATE_AVAILABLE:
                output.setText(getResources().getString(R.string.update_check_update_available, newestVersion));
                break;
        }
    }

    public void updateButtonClicked(View view) {
        switch (updateCheckState) {
            case BUSY:
                /* Do nothing. */
                break;
            case UPDATE_AVAILABLE:
                Manifest.Release latestRelease = updateCheckResult.getLatestRelease();
                Uri uri = URLs.toUri(latestRelease.getFileURL());
                String basename = uri.getLastPathSegment();
                if (basename == null) {
                    basename = updateCheckResult.getName().toLowerCase() + "-" + latestRelease.getVersion() + ".apk";
                }
                DownloadManager mgr = (DownloadManager) getSystemService(DOWNLOAD_SERVICE);
                DownloadManager.Request req = new DownloadManager.Request(uri);
                req.setMimeType("application/vnd.android.package-archive");
                try {
                    // Using private storage did not work in at least one test case (perhaps because of permission
                    // issues).
                    req.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, basename);
                } catch (IllegalStateException exc) {
                    Toast.makeText(this, R.string.update_download_failed_no_storage, Toast.LENGTH_SHORT).show();
                    return;
                }
                req.setDescription(getString(R.string.update_download_title, getString(R.string.app_name),
                        latestRelease.getVersion()));
                req.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
                mgr.enqueue(req);
                Toast.makeText(this, R.string.update_download_instructions, Toast.LENGTH_LONG).show();
                break;
            default:
                setUpdateCheckState(UpdateCheckState.BUSY);
                ManifestDownloader.download(URLs.toURL(URLs.getUpdateManifest()), new ManifestDownloader.Callback() {
                    @Override
                    public void downloadFinished(final Manifest result) {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                updateCheckFinished(result);
                            }
                        });
                    }

                    @Override
                    public void downloadFailed(final Throwable error) {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                updateCheckFailed(error);
                            }
                        });
                    }
                });
                break;
        }
    }

    private void updateCheckFinished(Manifest result) {
        updateCheckResult = result;
        if (result.getVersion().compareTo(Version.getCurrentAppVersion()) > 0) {
            setUpdateCheckState(UpdateCheckState.UPDATE_AVAILABLE);
        } else {
            setUpdateCheckState(UpdateCheckState.NO_RESULT);
        }
    }

    private void updateCheckFailed(Throwable error) {
        setUpdateCheckState(UpdateCheckState.FAILED);
        Log.w(UPDATE_CHECK_TAG, "Could not download update manifest: " + error, error);
    }

}

File: app/src/main/java/io/euphoria/xkcd/app/AboutActivity.kt
package io.euphoria.xkcd.app

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import io.euphoria.xkcd.app.updates.Manifest
import io.euphoria.xkcd.app.updates.ManifestDownloader
import io.euphoria.xkcd.app.updates.Version
import java.net.URL

class AboutActivity : Activity() {
    private enum class UpdateCheckState {
        BLANK, BUSY, FAILED, NO_RESULT, UPDATE_AVAILABLE
    }

    private lateinit var updateCheckState: UpdateCheckState
    private var updateCheckResult: Manifest? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.activity_about)
        updateCheckState = UpdateCheckState.BLANK
        val version: TextView = findViewById(R.id.about_version)
        version.text = resources.getString(
            R.string.version_line,
            BuildConfig.VERSION_NAME,
            resources.getString(R.string.app_version_label)
        )
    }

    private fun setUpdateCheckState(newState: UpdateCheckState) {
        updateCheckState = newState
        val newestVersion: Version? = if (updateCheckResult != null) {
            updateCheckResult!!.version
        } else {
            null
        }
        val button: Button =
            findViewById(R.id.about_updateCheck)
        val outputLine: ViewGroup = findViewById(R.id.about_updateCheckLine)
        val progress: ProgressBar = findViewById(R.id.about_updateCheckProgress)
        val output: TextView = findViewById(R.id.about_updateCheckResult)
        when (newState) {
            UpdateCheckState.BLANK -> {
                button.isEnabled = true
                outputLine.visibility = View.GONE
                progress.visibility = View.GONE
            }
            UpdateCheckState.BUSY -> {
                button.isEnabled = false
                outputLine.visibility = View.VISIBLE
                progress.visibility = View.VISIBLE
            }
            UpdateCheckState.FAILED, UpdateCheckState.NO_RESULT, UpdateCheckState.UPDATE_AVAILABLE -> {
                button.isEnabled = true
                outputLine.visibility = View.VISIBLE
                progress.visibility = View.GONE
            }
        }
        if (newState == UpdateCheckState.UPDATE_AVAILABLE) {
            button.text = resources.getString(R.string.update_check_download, newestVersion)
        } else {
            button.setText(R.string.update_check)
        }
        when (newState) {
            UpdateCheckState.BUSY -> output.setText(R.string.update_check_busy)
            UpdateCheckState.FAILED -> output.setText(R.string.update_check_failed)
            UpdateCheckState.NO_RESULT -> output.setText(R.string.update_check_no_result)
            UpdateCheckState.UPDATE_AVAILABLE -> output.text = resources.getString(
                R.string.update_check_update_available,
                newestVersion
            )
            UpdateCheckState.BLANK -> {/* this is a dummy state -> do nothing */}
        }
    }

    fun updateButtonClicked(view: View?) {
        when (updateCheckState) {
            UpdateCheckState.BUSY -> {
            }
            UpdateCheckState.UPDATE_AVAILABLE -> {
                val url: URL? = updateCheckResult!!.latestRelease?.fileURL
                val intent: Intent = Intent(Intent.ACTION_VIEW, URLs.toUri(url))
                if (packageManager.resolveActivity(
                        intent,
                        PackageManager.MATCH_DEFAULT_ONLY
                    ) == null
                ) {
                    Toast.makeText(this, R.string.update_check_open_failed, Toast.LENGTH_SHORT)
                        .show()
                    return
                }
                startActivity(intent)
            }
            else -> {
                setUpdateCheckState(UpdateCheckState.BUSY)
                ManifestDownloader.download(
                    URLs.toURL(URLs.updateManifest),
                    object : ManifestDownloader.Callback {
                        override fun downloadFinished(result: Manifest) {
                            runOnUiThread { updateCheckFinished(result) }
                        }

                        override fun downloadFailed(error: Throwable) {
                            runOnUiThread { updateCheckFailed(error) }
                        }
                    })
            }
        }
    }

    private fun updateCheckFinished(result: Manifest) {
        updateCheckResult = result
        if (result.version > Version.Companion.currentAppVersion
        ) {
            setUpdateCheckState(UpdateCheckState.UPDATE_AVAILABLE)
        } else {
            setUpdateCheckState(UpdateCheckState.NO_RESULT)
        }
    }

    private fun updateCheckFailed(error: Throwable) {
        setUpdateCheckState(UpdateCheckState.FAILED)
        Log.w(
            UPDATE_CHECK_TAG,
            "Could not download update manifest: $error",
            error
        )
    }

    companion object {
        private const val UPDATE_CHECK_TAG: String = "UpdateCheck"
    }
}
